### ISSUE TITLE
Running the code with pre-train step gives lower performance

### ISSUE THREAD

**Issue by gayanku** [nairouz/R-GAE#5]
Running the pre-train code results in lower accuracy.

Running `python R-GMM-VGAE/main_cora.py`  with the _supplied_ model.pk gives
100% 199/200 [01:01<00:00,  3.70it/s]ACC=0.750369, f1_macro=0.729495, precision_macro=0.750872, recall_macro=0.746598, f1_micro=0.750369, precision_micro=0.750369, recall_micro=0.750369, NMI=0.554135, ADJ_RAND_SCORE=0.551637

However, if I try to do my own pre-training  ( by adding 
`def pretrain(self, adj, features, adj_label, y, weight_tensor, norm, epochs, lr, save_path, dataset):
        ........
        torch.save( self.state_dict(), save_path + dataset + '/pretrain/model.pk' )`

I seem to be getting lower performance as shown below.
Loss=14.6272
100% 199/200 [01:10<00:00,  3.67it/s]ACC=0.658050, f1_macro=0.617192, precision_macro=0.625855, recall_macro=0.623322, f1_micro=0.658050, precision_micro=0.658050, recall_micro=0.658050, NMI=0.488975, ADJ_RAND_SCORE=0.465233

Whats the process to produce my own model.pk ?

**Comment by nairouz** [comment]
The pretraining phase is very important for reproducing the article results. So, I generally select the best accuracy during pretraining, similar to previous methods. For Cora, I get an accuracy around 69% as a pretraining result. You may have to repeat the pretraining phase 2 or  3 times.  I also make sure that R-GMM-VGAE and GMM-VGAE use the same pretraining file for fairness of comparison.

```
if acc > acc_best:
   acc_best = acc
   torch.save(self.state_dict(), save_path + dataset + '/pretrain/model_' + str(self.m) + '.pk')
```

**Comment by gayanku** [comment]
Thank you for the clarification.
I am still a bit confused how you would compute the acc. DO you use the labels? Wouldn't that mean you need annotations to obtain the optimal model during pre train?
Perhaps you can share your full pretrain model creation code. This may clarify the approach.

**Comment by nairouz** [comment]
Previous methods use this technique to select the best pretraining results, so it was necessary to use the same technique to reproduce the results of GMM-VGAE for example. Added to that, the labels are not used for performing pretraining, which is vanilla reconstruction. But, we can use the labels to select when to stop the pretraining phase, instead of repeating the pretraining phase several times and taking the best result. Nothing wrong with that. This technique is commonplace and is used by most if not all graph auto-encoder models that I have seen.

Furthermore, the aim of this work is not to beat the sota results. It is rather to show  the trade-off between FR and FD, and the experiments are conducted in a way that both models (R-GMM-VGAE and GMM-VGAE) start from the same pretraining level. I am already sharing the pretraining code, you can add the provided two lines to save the best pretraining weights, that's all.
